libcache/cache.c

```diff
@@ -0,0 +1,696 @@
+					cache = NULL;
+				}
+			}
+			else {
+				free(cache);
+				cache = NULL;
+			}
+		}
+	}
+
+	return cache;
+}
+
+
+static uint64_t cache_computeAddr(const cachectx_t *cache, uint64_t tag, uint64_t setIndex)
+{
+	uint64_t temp = (tag << cache->setBitsNum) | setIndex;
+	return temp <<= cache->offBitsNum;
+}
+
+
+static ssize_t cache_flushLine(cachectx_t *cache, cacheline_t *linePtr, uint64_t addr)
+{
+	ssize_t written = 0, position = 0;
+	size_t left = cache->lineSize;
+
+	if (linePtr != NULL && IS_VALID(linePtr->flags) && IS_DIRTY(linePtr->flags)) {
+		while (left > 0) {
+			written = cache->writeCb(addr, (unsigned char *)linePtr->data + position, left);
+
+			if (written <= 0) {
+				position = -EIO;
+				break;
+			}
+
+			left -= written;
+			addr += written;
+			position += written;
+		}
+
+		if (position == (ssize_t)cache->lineSize) {
+			CLEAR_DIRTY(linePtr->flags);
+		}
+	}
+
+	return position;
+}
+
+
+int cache_deinit(cachectx_t *cache)
+{
+	int i, j, ret = EOK, err;
+	uint64_t addr;
+	ssize_t written;
+	cacheline_t *linePtr;
+
+	if (cache == NULL) {
+		ret = -EINVAL;
+	}
+	else {
+		for (i = 0; i < cache->numSets && ret != -EIO; ++i) {
+			for (j = 0; j < LIBCACHE_NUM_WAYS; ++j) {
+				linePtr = &(cache->sets[i].lines[j]);
+
+				if (IS_VALID(linePtr->flags)) {
+					if (IS_DIRTY(linePtr->flags)) {
+						addr = cache_computeAddr(cache, linePtr->tag, i);
+						written = cache_flushLine(cache, linePtr, addr);
+
+						if (written < (ssize_t)cache->lineSize) {
+							ret = -EIO;
+							break;
+						}
+					}
+
+					free(linePtr->data);
+					CLEAR_VALID(linePtr->flags);
+				}
+			}
+		}
+
+		if (ret != -EIO) {
+			err = resourceDestroy(cache->lock);
+			if (err < 0) {
+				ret = err;
+			}
+			else {
+				free(cache->sets);
+				free(cache);
+			}
+		}
+	}
+
+	return ret;
+}
+
+
+static int cache_compareTags(const void *lhs, const void *rhs)
+{
+	int ret = 0;
+
+	const cacheline_t **lhsLine = (const cacheline_t **)lhs;
+	const cacheline_t **rhsLine = (const cacheline_t **)rhs;
+
+	if (*lhsLine == NULL && *rhsLine == NULL) {
+		ret = 0;
+	}
+	else if (*lhsLine == NULL && *rhsLine != NULL) {
+		ret = -1;
+	}
+	else if (*lhsLine != NULL && *rhsLine == NULL) {
+		ret = 1;
+	}
+	else {
+		if ((*lhsLine)->tag < (*rhsLine)->tag) {
+			ret = -1;
+		}
+		else if ((*lhsLine)->tag == (*rhsLine)->tag) {
+			ret = 0;
+		}
+		else {
+			ret = 1;
+		}
+	}
+
+	return ret;
+}
+
+
+static ssize_t cache_executePolicy(cachectx_t *cache, cacheline_t *linePtr, uint64_t addr, int policy)
+{
+	ssize_t written = cache->lineSize;
+
+	if (policy == 1) {
+		written = cache_flushLine(cache, linePtr, addr);
+		if (written <= 0) {
+			written = -EIO;
+		}
+	}
+
+	return written;
+}
+
+
+static cacheline_t *cache_allocateLine(cachectx_t *cache, const uint64_t setIndex, const uint64_t tag)
+{
+	int i, lineIndex;
+	uint64_t addr = 0;
+	cacheline_t *temp = NULL, *linePtr = NULL;
+	unsigned char flags = 0;
+	void *data;
+	ssize_t written = cache->lineSize;
```

Comment:
This one is somewhat misleading - you use this as de facto error flag. This makes last if (if (written == (ssize_t)cache->lineSize) {) somewhat misterious, because you check there if there was an error during flushLine.

Reply:
Fortunately, this is no longer an issue as the MISRA rule 15.5 was partially abandoned in ``cache_allocateLine`` function to enhance code readability.